mautic/mautic
# SMS support for received & read & failed stats

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:
This pull request introduces support for tracking delivered, read, and failed statistics in SMS messages. By default, it should function as it did before. However, it adds new columns to the sms_message and sms_messages tables. 

Any SMS transport integration can enhance the transport by incorporating additional interfaces to enable these statistics in the user interface. The transport itself should be responsible for managing the incrementing of each column's count. In the future, we may consider adding some services to simplify this process.

In another PR we can bring some helpers to easier manage data for transport.




#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2.  Run migration 
3.  Check If everything run as before  without changes
4.  Then dev during the code review can enabled Twilio with fake authorization data an add to Twilio transport new interfaces:

`class TwilioTransport implements TransportInterface, DeliverySupportInterface, ReadSupportInterface, FailedSupportInterface`


![image](https://github.com/mautic/mautic/assets/462477/2206f977-6dc4-456f-91c2-93fe947fdd32)

5. Then create SMS message
6. See on list page the new stats column
7. See on detail page new line graph options

![image](https://github.com/mautic/mautic/assets/462477/653c7ed8-e05e-438e-b6d9-aa5abe3a20b1)

![image](https://github.com/mautic/mautic/assets/462477/34e06450-1bb7-472a-b498-553eae35f2d2)
